app/src/main/java/de/westnordost/streetcomplete/quests/LastPickedValuesStore.kt

```diff
@@ -30,29 +30,38 @@ class LastPickedValuesStore<T : Any>(
 
 private const val MAX_ENTRIES = 100
 
-/* In the first `historyCount` items, return the `count` most-common non-null items, in order.
- * If the first item is not included (and is not null), it replaces the last of the common items.
- * If fewer than `count` unique items are found, continue counting items until that many are found,
- * or the end of the sequence is reached.
+/* Returns the `target` most-common non-null items in the first `historyCount`
+ *  items of the sequence, in their original order.
+ * If there are fewer than `target` unique items, continues counting items
+ *  until that many are found, or the end of the sequence is reached.
+ * If the first item (i.e. the most recent one) is not null, it is always
+ *  included, displacing the least-common of the other items if necessary.
  */
-fun <T : Any> Sequence<T?>.mostCommonWithin(count: Int, historyCount: Int): Sequence<T> {
-    val counts = this.countUniqueNonNull(historyCount, count)
-    val top = counts.keys.sortedByDescending { counts.get(it) }
+fun <T : Any> Sequence<T?>.mostCommonWithin(target: Int, historyCount: Int): Sequence<T> {
+    val counts = this.countUniqueNonNull(target, historyCount)
+    val top = counts.keys.sortedByDescending { counts[it]!!.count }.take(target)
     val latest = this.take(1).filterNotNull()
-    val items = (latest + top).distinct().take(count)
-    return items.sortedByDescending { counts.get(it) }
+    val items = (latest + top).distinct().take(target)
+    return items.sortedBy { counts[it]!!.indexOfFirst }
 }
 
+private data class ItemStats(val indexOfFirst: Int, var count: Int = 0)
+
 // Counts at least the first `minItems`, keeps going until it finds at least `target` unique values
-private fun <T : Any> Sequence<T?>.countUniqueNonNull(minItems: Int, target: Int): Map<T, Int> {
-    val counts = mutableMapOf<T, Int>()
-    val items = takeAtLeastWhile(minItems) { counts.size < target }.filterNotNull()
-    return items.groupingBy { it }.eachCountTo(counts)
-}
+private fun <T : Any> Sequence<T?>.countUniqueNonNull(target: Int, minItems: Int): Map<T, ItemStats> {
+    val counts = mutableMapOf<T, ItemStats>()
 
-// Take at least `count` elements, then continue until `predicate` returns false
-private fun <T> Sequence<T>.takeAtLeastWhile(count: Int, predicate: (T) -> Boolean): Sequence<T> =
-    withIndex().takeWhile{ (i, t) -> i < count || predicate(t) }.map { it.value }
+    // Because this is a sequence, the chain of calls is applied to items one at a time.
+    // This means the `getOrPut` for one item affects the `counts.size` for the next item
+    this.withIndex()
+        .filter { it.value != null }
+        .takeWhile { it.index < minItems || counts.size < target }
+        .forEach {
+            counts.getOrPut(it.value!!) { ItemStats(it.index) }.count++
+        }
+
+    return counts
+}
```

Comment:
Reply to https://github.com/streetcomplete/StreetComplete/pull/3564#discussion_r770959686:

Why not use a `for` loop then? We're not really making use of the functional style in this case, and it's looking rather imperative anyway. A `for` loop is easily understandable, even if you don't know the advantages and intricacies of sequences.

```kt
val counts = mutableMapOf<T, ItemStats>()

// maybe there is a nicer indexed for...in loop style
for (var index = 0; index < this.size; index++) {
    if (index >= minItems && counts.size >= target) break

    if (this[index] != null) {
        counts.getOrPut(this[index]) { ItemStats(index) }.count++
    }
}

return counts
```

Reply:
1. You can't index into a Sequence, so this exact version wouldn't work.
2. Sequences are important because it's sequences the whole way down. Values are stored as a comma-separated string in the SharedPreferences. We retrieve that, and then split directly to a sequence— laziness here means that if we only end up checking the first 30 values, we'll only do 30 splits. Admittedly, I didn't profile here to make sure this optimization is necessary.

I think I may be able to achieve the same effect using a while loop and an iterator. I'll try that.

Reply:
I interpret the 👍 by Flo as "Ok, resolved then".

Does this mean that Flo's code review is done and resolved?

Reply:
I interpreted it as "this will be resolved once you make the change".

Reply:
Turns out a `for` loop actually works great.